core/drivers/regulator/stm32mp1_regulator_iod.c

```diff
@@ -0,0 +1,296 @@
+ * entering suspend state and restore the configuration when resuming.
+ */
+static TEE_Result iod_pm(enum pm_op op, unsigned int pm_hint __unused,
+			 const struct pm_callback_handle *hdl)
+{
+	struct regulator *regu = hdl->handle;
+	struct iod_regul *iod = regu->priv;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	assert(op == PM_OP_SUSPEND || op == PM_OP_RESUME);
+
+	if (op == PM_OP_SUSPEND) {
+		FMSG("%s: suspend", regulator_name(regu));
+
+		res = iod_get_state(regu, &iod->suspend_state);
+		if (res)
+			return res;
+
+		res = iod_get_voltage(regu, &iod->suspend_level_uv);
+		if (res)
+			return res;
+
+		stm32mp_set_hslv_state(iod->hslv_id, false);
+	} else {
+		FMSG("%s: resume", regulator_name(regu));
+
+		res = iod_set_voltage(regu, iod->suspend_level_uv);
+		if (res)
+			return res;
+
+		res = iod_set_state(regu, iod->suspend_state);
+		if (res)
+			return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result iod_supplied_init(struct regulator *regu,
+				    const void *fdt __unused, int node __unused)
+{
+	struct iod_regul *iod = regu->priv;
+	int index = iod - iod_regulator_priv;
+
+	assert(index >= 0 && index < IOD_REGU_COUNT);
+
+	if (regulator_get_voltage(regu) < IO_VOLTAGE_THRESHOLD_UV)
+		stm32mp_set_hslv_state(iod->hslv_id, true);
+
+	/* Save regulator reference */
+	iod_regulator[index] = regu;
+
+	register_pm_driver_cb(iod_pm, regu, "iod-regulator");
+
+	FMSG("IOD regulator %s intiialized", regulator_name(regu));
+
+	return TEE_SUCCESS;
+}
+
+static const struct regulator_ops iod_ops = {
+	.set_state = iod_set_state,
+	.get_state = iod_get_state,
+	.set_voltage = iod_set_voltage,
+	.get_voltage = iod_get_voltage,
+	.supported_voltages = iod_list_voltages,
+	.supplied_init = iod_supplied_init,
+};
+
+#define DEFINE_REG(_id, _name, _supply_name) { \
+	.name = (_name), \
+	.ops = &iod_ops, \
+	.priv = iod_regulator_priv + (_id), \
+	.supply_name = (_supply_name), \
+}
+
+static struct regu_dt_desc iod_regul_desc[IOD_REGU_COUNT] = {
+	[IOD_SDMMC1] = DEFINE_REG(IOD_SDMMC1, "sdmmc1_io", "vddsd1"),
+	[IOD_SDMMC2] = DEFINE_REG(IOD_SDMMC2, "sdmmc2_io", "vddsd2"),
+};
+
+static TEE_Result iod_regulator_probe(const void *fdt, int node,
+				      const void *compat_data __unused)
+{
+	const char *node_name = NULL;
+	size_t i = 0;
+
+	node_name = fdt_get_name(fdt, node, NULL);
+
+	FMSG("iod probe node '%s'", node_name);
+
+	/* Look up matching regulator name defined in SoC DTSI file */
+	for (i = 0; i < IOD_REGU_COUNT; i++)
+		if (!strcmp(iod_regul_desc[i].name, node_name))
+			break;
+
+	if (i == IOD_REGU_COUNT) {
+		EMSG("Unexpected IO domain node name '%s'", node_name);
+		return TEE_ERROR_GENERIC;
+	}
+
+	return regulator_dt_register(fdt, node, node, iod_regul_desc + i);
+}
+
+static const struct dt_device_match iod_regulator_match_table[] = {
+	{ .compatible = "st,stm32mp13-iod" },
+	{ }
+};
+
+DEFINE_DT_DRIVER(stm32mp1_regulator_iod_dt_driver) = {
+	.name = "stm32mp1-iod-regulator",
```

Comment:
This driver is only supported on STM32MP13x.

Shall we rename it or are we anticipating changes on STM32MP15?

Reply:
> STM32MP15 does have have needs runtime voltage level re-configurable IO domains. I'll rename the driver to "stm32mp13-iod-regulator".

I guess the file name itself can be changed as well

Reply:
make sense, i'll do.